### PR TITLE
[dbnode] Bootstrap followups to ensure bootstrapped data is not for persisted windows

### PR DESCRIPTION
**Pull request by robskillington**
**What this PR does / why we need it**:

This adds back previous sanity checks.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/dbnode/storage/types.go [hunk 1]
```diff
@@ -338,6 +338,11 @@ type databaseNamespace interface {
 		opts block.FetchBlocksMetadataOptions,
 	) (block.FetchBlocksMetadataResults, PageToken, error)
 
+	// PrepareBootstrap prepares the namespace for bootstrapping by ensuring
+	// it's shards knows which flushed files reside on disk, so that calls
```
Comment (arnikola) [inline src/dbnode/storage/types.go]: nit;`its shards know...`

File: src/dbnode/storage/types.go [hunk 2]
```diff
@@ -488,6 +493,10 @@ type databaseShard interface {
 		opts block.FetchBlocksMetadataOptions,
 	) (block.FetchBlocksMetadataResults, PageToken, error)
 
+	// PrepareBootstrap prepares the shard for bootstrapping by ensuring
+	// it knows which flushed files reside on disk.
+	PrepareBootstrap() error
```
Comment (arnikola) [inline src/dbnode/storage/types.go]: test

File: src/dbnode/storage/shard.go [hunk 3]
```diff
@@ -1804,6 +1804,22 @@ func (s *dbShard) FetchBlocksMetadataV2(
...
+func (s *dbShard) PrepareBootstrap() error {
+	s.flushState.RLock()
+	flushStateBootstrapped := s.flushState.bootstrapped
+	s.flushState.RUnlock()
+	if flushStateBootstrapped {
+		return nil // Already prepared.
+	}
+
+	// Iterate flushed time ranges to determine which blocks are retrievable.
+	// NB(r): This must be done before bootstrap since during bootstrapping
```
Comment (arnikola) [inline src/dbnode/storage/shard.go]: Is it worth adding some sort of flag that will ensure this is the case, or maybe setting `s.flushState` to some sort of `bootstrappingPrepared` state?

File: src/dbnode/storage/shard.go [hunk 4]
```diff
@@ -2270,18 +2283,20 @@ func (s *dbShard) Snapshot(
 
 func (s *dbShard) FlushState(blockStart time.Time) (fileOpState, error) {
 	s.flushState.RLock()
-	defer s.flushState.RUnlock()
 	if !s.flushState.bootstrapped {
+		s.flushState.RUnlock()
```
Comment (arnikola) [inline src/dbnode/storage/shard.go]: Is this a hot enough path that this is necessary? Otherwise with multiple paths defer may be safer here

File: src/dbnode/storage/series/series.go [hunk 5]
```diff
@@ -288,6 +289,19 @@ func (s *dbSeries) Write(
...
+	if wOpts.BootstrapWrite {
+		at := timestamp.Truncate(s.opts.RetentionOptions().BlockSize())
+		alreadyExists, err := s.blockRetriever.IsBlockRetrievable(at)
+		if err != nil {
+			return false, fmt.Errorf(
+				"error checking block retrievable for bootstrap write: %v", err)
+		}
+		if alreadyExists {
+			return false, fmt.Errorf(
+				"bootstrap write for block is retrievable: block_start=%s", at)
```
Comment (arnikola) [inline src/dbnode/storage/series/series.go]: meganit: this message doesn't read like an error, maybe reword it a little to `cannot write block at %s: already exists` or something along those lines?

File: src/dbnode/storage/series/series.go [hunk 6]
```diff
@@ -391,6 +405,19 @@ func (s *dbSeries) LoadBlock(
...
+	switch writeType {
+	case WarmWrite:
+		at := block.StartTime()
+		alreadyExists, err := s.blockRetriever.IsBlockRetrievable(at)
+		if err != nil {
+			return fmt.Errorf("error checking block retrievable: %v", err)
+		}
+		if alreadyExists {
+			return fmt.Errorf(
+				"block load as warm write is retrievable: block_start=%s", at)
```
Comment (arnikola) [inline src/dbnode/storage/series/series.go]: meganit: similarily reword error here maybe?

File: src/dbnode/storage/series/series.go [hunk 7]
```diff
@@ -288,6 +289,19 @@ func (s *dbSeries) Write(
 	annotation []byte,
 	wOpts WriteOptions,
 ) (bool, error) {
+	if wOpts.BootstrapWrite {
+		at := timestamp.Truncate(s.opts.RetentionOptions().BlockSize())
+		alreadyExists, err := s.blockRetriever.IsBlockRetrievable(at)
+		if err != nil {
+			return false, fmt.Errorf(
+				"error checking block retrievable for bootstrap write: %v", err)
```
Comment (arnikola) [inline src/dbnode/storage/series/series.go]: Worth adding `at` to message?

File: src/dbnode/storage/series/buffer.go [hunk 8]
```diff
@@ -273,10 +273,10 @@ func (b *dbBuffer) Write(
 	)
 	switch {
 	case wOpts.BootstrapWrite:
+		// Bootstrap writes are always warm writes, validated
+		// that safe by the series.
```
Comment (arnikola) [inline src/dbnode/storage/series/buffer.go]: nit: maybe something along the lines of  
```
Bootstrap writes are always warm writes
NB: callers must validate that the block does not reside on disk
```

File: src/dbnode/storage/namespace.go [hunk 9]
```diff
@@ -750,6 +750,37 @@ func (n *dbNamespace) AggregateQuery(
...
+	for _, shard := range shards {
+		shard := shard
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			err := shard.PrepareBootstrap()
+			if err != nil {
+				multiErrLock.Lock()
+				multiErr = multiErr.Add(err)
```
Comment (arnikola) [inline src/dbnode/storage/namespace.go]: May be worth logging here to make bootstrap debugging a little easier?

File: src/dbnode/storage/bootstrap.go [hunk 10]
```diff
@@ -228,16 +233,54 @@ func (m *bootstrapManager) bootstrap() error {
...
+		go func() {
+			defer prepareWg.Done()
+
+			shards, err := namespace.PrepareBootstrap()
+
+			prepareLock.Lock()
+			defer prepareLock.Unlock()
+
+			if err != nil {
+				prepareMultiErr = prepareMultiErr.Add(err)
```
Comment (arnikola) [inline src/dbnode/storage/bootstrap.go]: nit: worth logging the error here?

File: src/dbnode/storage/bootstrap.go [hunk 11]
```diff
@@ -228,16 +233,54 @@ func (m *bootstrapManager) bootstrap() error {
...
+	for i, namespace := range namespaces {
+		i, namespace := i, namespace
+		prepareWg.Add(1)
+		go func() {
+			defer prepareWg.Done()
+
+			shards, err := namespace.PrepareBootstrap()
+
+			prepareLock.Lock()
+			defer prepareLock.Unlock()
```
Comment (arnikola) [inline src/dbnode/storage/bootstrap.go]: nit: can combine into one?
```
defer func() {
 prepareWg.Done()
 prepareLock.Unlock()
}()
```

File: src/dbnode/storage/series/series.go [hunk 12]
```diff
@@ -391,6 +405,19 @@ func (s *dbSeries) LoadBlock(
 	block block.DatabaseBlock,
 	writeType WriteType,
 ) error {
+	switch writeType {
+	case WarmWrite:
+		at := block.StartTime()
+		alreadyExists, err := s.blockRetriever.IsBlockRetrievable(at)
```
Comment (arnikola) [inline src/dbnode/storage/series/series.go]: This seems to be hitting nil references in tests, it looks like `NewDatabaseSeries` does not set a `blockRetriever`

File: src/dbnode/storage/mediator.go [hunk 13]
```diff
@@ -194,6 +194,7 @@ func (m *mediator) ongoingFilesystemProcesses() {
 			mediatorTime, err := m.mediatorTimeBarrier.fsProcessesWait()
 			if err != nil {
 				log.Error("error within ongoingFilesystemProcesses waiting for next mediatorTime", zap.Error(err))
+				continue
```
Comment (arnikola) [inline src/dbnode/storage/mediator.go]: Has this just been broken for ages?

File: src/dbnode/storage/bootstrap_test.go [hunk 14]
```diff
@@ -64,7 +64,7 @@ func TestDatabaseBootstrapWithBootstrapError(t *testing.T) {
 	bsm.sleepFn = func(time.Duration) {}
 
 	gomock.InOrder(
-		ns.EXPECT().GetOwnedShards().Return([]databaseShard{}),
+		ns.EXPECT().PrepareBootstrap().Return([]databaseShard{}, nil),
```
Comment (arnikola) [inline src/dbnode/storage/bootstrap_test.go]: Can't comment on the line unfortunately, but maybe update the Controller to be an `xtest.NewController(t)`?

File: src/dbnode/storage/shard_test.go [hunk 15]
```diff
@@ -811,7 +811,11 @@ func addTestSeries(shard *dbShard, id ident.ID) series.DatabaseSeries {
 }
 
 func addTestSeriesWithCount(shard *dbShard, id ident.ID, count int32) series.DatabaseSeries {
-	seriesEntry := series.NewDatabaseSeries(id, ident.Tags{}, 1, shard.seriesOpts)
+	seriesEntry := series.NewDatabaseSeries(series.DatabaseSeriesOptions{
```
Comment (arnikola) [inline src/dbnode/storage/shard_test.go]: General style question: Should we keep `FooOptions` nomenclature specifically for interfaces with getters and setters, or just for collections of correlated values?

File: src/dbnode/storage/series/types.go [hunk 16]
```diff
@@ -133,15 +144,7 @@ type DatabaseSeries interface {
...
-	Reset(
-		id ident.ID,
-		tags ident.Tags,
-		uniqueIndex uint64,
-		blockRetriever QueryableBlockRetriever,
-		onRetrieveBlock block.OnRetrieveBlock,
-		onEvictedFromWiredList block.OnEvictedFromWiredList,
-		opts Options,
-	)
+	Reset(opts DatabaseSeriesOptions)
```
Comment (arnikola) [inline src/dbnode/storage/series/types.go]: Would it be a good idea to change Reset to return an error, e.g. if some of the necessary fields in `DatabaseSeriesOptions` are unset?

File: src/dbnode/storage/shard.go [hunk 17]
```diff
@@ -1804,6 +1811,22 @@ func (s *dbShard) FetchBlocksMetadataV2(
...
+	if flushStateBootstrapped {
+		return nil // Already prepared.
+	}
+
+	// Iterate flushed time ranges to determine which blocks are retrievable.
+	// NB(r): This must be done before bootstrap since during bootstrapping
+	// series will load blocks into series with series.LoadBlock(...) which
+	// needs to ask the shard whether certain time windows have been flushed or
+	// not.
+	return s.bootstrapFlushStates()
```
Comment (arnikola) [inline src/dbnode/storage/shard.go]: `bootstrapFlushStates` again checks `s.flushState.bootstrapped`, potentially racey if something happens between the checks?

File: src/dbnode/storage/shard.go (hunk 4 above)
Comment (robskillington) [reply src/dbnode/storage/shard.go]: Will update so only one RUnlock()

File: src/dbnode/storage/shard.go (hunk 3 above)
Comment (robskillington) [reply src/dbnode/storage/shard.go]: Yup I've changed `bootstrapped` bool to `initialized` bool to track this.

File: src/dbnode/storage/types.go (hunk 1 above)
Comment (robskillington) [reply src/dbnode/storage/types.go]: Sounds good.

File: src/dbnode/storage/namespace.go (hunk 9 above)
Comment (robskillington) [reply src/dbnode/storage/namespace.go]: I do that below with I believe with bootstrap prepare failed?  I can wrap the error potentially.

File: src/dbnode/storage/bootstrap.go (hunk 10 above)
Comment (robskillington) [reply src/dbnode/storage/bootstrap.go]: I do that elsewhere with I believe with `"bootstrap prepare failed"` - to avoid log spam was thinking just doing this once.

File: src/dbnode/storage/bootstrap.go (hunk 11 above)
Comment (robskillington) [reply src/dbnode/storage/bootstrap.go]: Sounds good.

File: src/dbnode/storage/series/series.go (hunk 12 above)
Comment (robskillington) [reply src/dbnode/storage/series/series.go]: Yup, I fixed this by updating all the tests with an opts struct too.

File: src/dbnode/storage/mediator.go (hunk 13 above)
Comment (robskillington) [reply src/dbnode/storage/mediator.go]: Yeah, looks like it.

File: src/dbnode/storage/series/types.go (hunk 16 above)
Comment (robskillington) [reply src/dbnode/storage/series/types.go]: Maybe we do this as a follow up?  It has wide reaching implications unfortunately due to the amount of callers.

File: src/dbnode/storage/shard_test.go (hunk 15 above)
Comment (robskillington) [reply src/dbnode/storage/shard_test.go]: My gut feel is to use it wherever it makes refactoring easy (look at how many call sites I had to update just to add an extra optional arg to `NewDatabaseSeries(...)`).

File: src/dbnode/storage/shard.go (hunk 17 above)
Comment (robskillington) [reply src/dbnode/storage/shard.go]: Yeah, I think this is ok though, it locks when it actually sets values.

File: src/dbnode/storage/series/buffer.go [hunk 18]
```diff
@@ -272,11 +272,6 @@ func (b *dbBuffer) Write(
 		writeType   WriteType
 	)
 	switch {
-	case wOpts.BootstrapWrite:
```
Comment (notbdu) [inline src/dbnode/storage/series/buffer.go]: Does this mean bootstrap writes will be classified as cold writes?

Comment (robskillington) [reply src/dbnode/storage/series/buffer.go]: I changed this so that basically whether the writes are for file volumes that already exist or not the bootstrap writes will be considered a cold write or not. 

This I believe reflects accurately what type of write it is when it's being accumulated from the commit log.
